streaming/format/base/file.py

```diff
@@ -0,0 +1,302 @@
+from streaming.format.base.phaser import Locality
+from streaming.format.canonical import canonicalize
+from streaming.stream.dir_conf import StreamDirConf
+
+__all__ = ['ShardFile']
+
+
+class ShardFile:
+    """Metadata for validating every phase of file across decompression and/or preparation.
+
+            *
+            |
+        Downloading
+            |
+          [zip]
+            |
+        Decompression
+            |
+          [raw]
+            |
+        Canonicalization
+            |
+          [can]
+            |
+        Eviction
+            |
+            v
+
+    Shard formats:
+      * MDS: zip -> raw (raw phase is performant, so no need for a can phase).
+      * JSONL: zip -> raw (").
+      * XSV: zip -> raw (").
+      * Parquet: raw -> can (using pre-existing third-party Parquet files, so can't compress, and
+        Parquet from-disk random access is bad enough to require converting to a usable format).
+
+    Notes:
+      * All shard files are required to have a raw phase.
+      * Every non-index file of a serialized Streaming dataset belongs to some Shard.
+      * Currently, all types of Shard are comprised of either one or two shard files.
+      * All files of a Shard must either jointly exist or not exist in order to be in a coherent
+        state.
+      * Files transition across their applicable phases unidirectionally forward.
+      * Files are stored officially, and therefore downloaded, in their first phase, and used in
+        their last phase.
+
+    Args:
+        stream (StreamDirConf): Link back up to the Stream that owns this shard, from which we get
+            arguments which are shared across all shards like remote/local paths. Avoids an import
+            cycle by Stream subclassing StreamDirConf.
+        zip_phase (ShardFilePhase, optional): Metadata for validating the compressed phase of the
+            file. Defaults to ``None``.
+        zip_algo (str, optional): Decompression algorithm, if zip is used. Defaults to ``None``.
+        raw_phase (ShardFilePhase): Metadata for validating the regular phase of the file.
+        can_algo (str, optional): Canonicalization algorithm, if can is used. Defaults to ``None``.
+        can_phase (ShardFilePhase, optional): Metadata for validating the canonicalized phase of
+            the file. Defaults to ``None``.
+    """
+
+    def __init__(
+        self,
+        *,
+        stream: StreamDirConf,
+        zip_phase: Optional[ShardFilePhase] = None,
+        zip_algo: Optional[str] = None,
+        raw_phase: ShardFilePhase,
+        can_algo: Optional[str] = None,
+        can_phase: Optional[ShardFilePhase] = None,
+    ) -> None:
+        self.stream = stream
+        self.zip_phase = zip_phase
+        self.zip_algo = zip_algo
+        self.raw_phase = raw_phase
+        self.can_algo = can_algo
+        self.can_phase = can_phase
+        self.phases = zip_phase, raw_phase, can_phase
```

Comment:
nit: cleaner if self.phases = (p for p in (zip_phase, raw_phase, can_phase) if p is not None). since ``if phase`` is performed several times in the code.

Reply:
Position is meaningful, None is a legitimate value

It’s not a list of the phases which are in use by this shard file instance, it’s a list of what this shard file instance has for each of the possible phases, in order